very weird warning when using same varname for const and let binding

### Code

```Rust

pub fn main() {
    const y: i32 = 4;
    let y: i32 = 3;
}
```

This will cause rustc to interpret the `y` as a `if let ` binding for some reason which seems quite surprising.
```
error[E0005]: refutable pattern in local binding
 --> src/main.rs:3:9
  |
3 |     let y: i32 = 3;
  |         ^ patterns `i32::MIN..=3_i32` and `5_i32..=i32::MAX` not covered
  |
  = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
  = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
  = note: the matched value is of type `i32`
help: you might want to use `if let` to ignore the variants that aren't matched
  |
3 |     if let y: i32 = 3 { todo!() };
  |     ++                +++++++++++

For more information about this error, try `rustc --explain E0005`.
```

When I replace `let` by `static`, I get a warning that seems to be much closer to reality:

```rust
pub fn main() {
    const y: i32 = 4;
    static y: i32 = 3;
}

```

````
error[E0428]: the name `y` is defined multiple times
 --> src/main.rs:3:5
  |
2 |     const y: i32 = 4;
  |     ----------------- previous definition of the value `y` here
3 |     static y: i32 = 3;
  |     ^^^^^^^^^^^^^^^^^^ `y` redefined here
  |
  = note: `y` must be defined only once in the value namespace of this block

For more information about this error, try `rustc --explain E0428`.
error: could not compile `playground` (bin "playground") due to previous error
````

IMO it would make sense to issue such warning for the first case as well, instead of the if-let/range diagnostic.
```

--- a/compiler/rustc_mir_build/src/thir/pattern/check_match.rs
+++ b/compiler/rustc_mir_build/src/thir/pattern/check_match.rs
@@ -443,7 +443,12 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
         let mut let_suggestion = None;
         let mut misc_suggestion = None;
         let mut interpreted_as_const = None;
-        if let PatKind::Constant { .. } = pat.kind
+
+        if let PatKind::Constant { .. }
+            | PatKind::AscribeUserType {
+                subpattern: box Pat { kind: PatKind::Constant { .. }, .. },
+                ..
+              } = pat.kind
             && let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(pat.span)
         {
             // If the pattern to match is an integer literal:

--- a/tests/ui/mir/issue-112269.rs
+++ b/tests/ui/mir/issue-112269.rs
@@ -1,0 +1,9 @@
+pub fn main() {
+    const x: i32 = 4;
+    let x: i32 = 3;
+    //~^ ERROR refutable pattern in local binding
+
+    const y: i32 = 3;
+    let y = 4;
+    //~^ ERROR refutable pattern in local binding
+}

--- a/tests/ui/mir/issue-112269.stderr
+++ b/tests/ui/mir/issue-112269.stderr
@@ -1,0 +1,31 @@
+error[E0005]: refutable pattern in local binding
+  --> $DIR/issue-112269.rs:3:9
+   |
+LL |     let x: i32 = 3;
+   |         ^
+   |         |
+   |         patterns `i32::MIN..=3_i32` and `5_i32..=i32::MAX` not covered
+   |         missing patterns are not covered because `x` is interpreted as a constant pattern, not a new variable
+   |         help: introduce a variable instead: `x_var`
+   |
+   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
+   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
+   = note: the matched value is of type `i32`
+
+error[E0005]: refutable pattern in local binding
+  --> $DIR/issue-112269.rs:7:9
+   |
+LL |     let y = 4;
+   |         ^
+   |         |
+   |         patterns `i32::MIN..=2_i32` and `4_i32..=i32::MAX` not covered
+   |         missing patterns are not covered because `y` is interpreted as a constant pattern, not a new variable
+   |         help: introduce a variable instead: `y_var`
+   |
+   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
+   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
+   = note: the matched value is of type `i32`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0005`.

